### PR TITLE
Improvements to language model support

### PR DESCRIPTION
**Pull request by ejnnr**
- Removes `HookedModel` and replaces it with a simpler approach based on pytorch forward hooks. The advantage is that we can directly use Huggingface or any other models with very little wrapper code.
- Adds an activation cache for `ActivationBasedDetector`, which makes it very easy to run multiple different detectors without constantly re-running the model. (Gathering the activations is _by far_ the most expensive part of e.g. training a Mahalanobis detector for LLMs.)
- Adds another demo notebook for quirky language model experiments. I might replace or update this though.

Tests are probably broken for now because I got rid of `HookedModel`, I'm guessing this will require a few minor fixes.

### CODE REVIEW

**Comment by ejnnr** [comment]
All tests work now except those related to abstractions. I started trying to fix the abstraction tests but there are a ton of places where we have to change names, and I think we probably want to rework how abstractions are done anyway (get rid of the "default abstraction" thing). Maybe I'll just rework abstractions in this PR, not sure yet.

**Comment by ejnnr** [comment]
Cleaned up some of the code, in particular the quirky LM task and required models/dataset is now defined in the main code based instead of just notebooks.
- [ ] Use that new definition in the demo notebook
- [ ] Replace the file paths for models with Huggingface names (after uploading them)
- [ ] Train new models on the customized datasets and get rid of the special case in `quirky_lm` (right now, the non-mixture models are trained on the original EleutherAI dataset with less data)

File: src/cupbearer/detectors/statistical/mahalanobis_detector.py [hunk 1]
```diff
@@ -1,19 +1,61 @@
...
 from cupbearer.detectors.statistical.statistical import (
     ActivationCovarianceBasedDetector,
 )
 
 
+def _pinv(C, rcond):
+    # Workaround for pinv not being supported on MPS
+    if C.device.type == "mps":
+        return torch.linalg.pinv(C.cpu(), rcond=rcond, hermitian=True).to(C.device)
+    return torch.linalg.pinv(C, rcond=rcond, hermitian=True)
```
Comment (VRehnberg) [inline src/cupbearer/detectors/statistical/mahalanobis_detector.py:15]: Alternatively https://lightning.ai/docs/pytorch/stable/accelerators/mps_basic.html#mps-training-basic or calculating pinv from eigh (if torch.linalg.eigh is supported on mps). But, I have no real objections against this either.

Comment (ejnnr) [reply src/cupbearer/detectors/statistical/mahalanobis_detector.py:15]: Probably we should be using https://pytorch.org/docs/stable/generated/torch.linalg.lstsq.html (or at least something other than inverting the matrix), will see if that works

File: src/cupbearer/detectors/activation_based.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+def model_checksum(model: torch.nn.Module) -> float:
+    """Hacky but fast way to get a checksum of the model parameters.
+    Just sums the absolute values.
+    """
+    device = next(model.parameters()).device
+    # Using float64 to get some more bits and make collisions less likely.
+    param_sum = torch.tensor(0.0, dtype=torch.float64, device=device)
+    for param in model.parameters():
+        param_sum += param.abs().sum()
+    return param_sum.item()
```
Comment (VRehnberg) [inline src/cupbearer/detectors/activation_based.py:24]: But the parameters are only 32 bit, right? You remove risk of overflows, but shouldn't collision be just as likely?

Also, I'm slightly concerned about floating point errors here, in case you cache the activations on one machine and then run on a different one with different instruction sets.

You could try:
```suggestion
    # Using int64 to get some more bits and make collisions less likely.
    param_sum = torch.tensor(0, dtype=torch.int64, device=device)
    for param in model.parameters():
        param_sum += param.view(torch.int32).to(torch.int64).sum()
    return param_sum.item()
```

File: src/cupbearer/detectors/statistical/helpers.py [hunk 3]
```diff
@@ -38,65 +38,60 @@ def batch_covariance(batches):
...
 def mahalanobis(
-    activations: dict[str, torch.Tensor],
-    means: dict[str, torch.Tensor],
-    inv_covariances: dict[str, torch.Tensor],
-    inv_diag_covariances: Optional[dict[str, torch.Tensor]] = None,
+    activation: torch.Tensor,
+    mean: torch.Tensor,
+    inv_covariance: torch.Tensor,
+    inv_diag_covariance: Optional[torch.Tensor] = None,
 ):
```
Comment (VRehnberg) [inline src/cupbearer/detectors/statistical/helpers.py:45]:
```suggestion
) -> torch.Tensor:
```

File: src/cupbearer/detectors/statistical/mahalanobis_detector.py [hunk 4]
```diff
@@ -6,30 +7,75 @@
...
+
+
+# From https://gist.github.com/chausies/011df759f167b17b5278264454fff379
+def norm_cdf(x):
+    return (1 + torch.erf(x / np.sqrt(2))) / 2
+
+
+def log_norm_cdf_helper(x):
+    a = 0.344
+    b = 5.334
```
Comment (VRehnberg) [inline src/cupbearer/detectors/statistical/mahalanobis_detector.py:24]: I get that the gist didn't have a source for these numbers, but they are still a bit suss for me. Also, precision is kinda low, though, probably doesn't matter as long as same numbers are used everytime.

Comment (VRehnberg) [reply src/cupbearer/detectors/statistical/mahalanobis_detector.py:24]: Never mind on the source, found it here from the comment on the gist https://en.wikipedia.org/wiki/Q-function#Bounds_and_approximations

File: src/cupbearer/detectors/statistical/statistical.py [hunk 5]
```diff
@@ -90,6 +103,42 @@ def batch_update(self, activations: dict[str, torch.Tensor]):
...
+
+        Should return a tensor of shape (batch,) with the anomaly scores.
+        """
+        pass
+
+    def layerwise_scores(self, batch):
+        activations = self.get_activations(batch)
+        batch_size = next(iter(activations.values())).shape[0]
+        activations = {
+            k: rearrange(v, "batch ... dim -> (batch ...) dim")
```
Comment (VRehnberg) [inline src/cupbearer/detectors/statistical/statistical.py:126]: So `v.flatten(end_dim=-2)`? I'm a bit unused to this notation.

File: src/cupbearer/detectors/statistical/statistical.py [hunk 6]
```diff
@@ -90,6 +103,42 @@ def batch_update(self, activations: dict[str, torch.Tensor]):
...
+            for k, v in activations.items()
+        }
+        scores = {
+            k: self._individual_layerwise_score(k, v) for k, v in activations.items()
+        }
+        scores = {
+            k: rearrange(
+                v,
+                "(batch independent_dims) -> batch independent_dims",
```
Comment (VRehnberg) [inline src/cupbearer/detectors/statistical/statistical.py:135]: Oh, and here it can infer size of independent dims? So basically `v.reshape(batch_size, -1)`

File: src/cupbearer/models/models.py [hunk 7]
```diff
@@ -31,19 +29,12 @@ def __init__(
...
             x = layer(x)
-            self.store(f"post_{name}", x)
         return x
 
-    @property
-    def default_names(self) -> list[str]:
-        return [
-            f"post_{name}" for name in self.layers.keys() if name.startswith("linear")
-        ]
-
```
Comment (VRehnberg) [inline src/cupbearer/models/models.py:44]: Good riddance.

File: src/cupbearer/utils/__init__.py [hunk 8]
```diff
@@ -36,6 +39,7 @@ def validate_and_convert_leaf(leaf):
     if isinstance(leaf, type):
         return TYPE_PREFIX + leaf.__module__ + "." + leaf.__name__
 
+    # TODO (erik): is any of this still necessary? torch.save should handle this I think
```
Comment (VRehnberg) [inline src/cupbearer/utils/__init__.py:42]: Good idea to add names to TODOs. I should start doing that as well.

File: tests/test_abstractions.py [hunk 9]
```diff
@@ -1,302 +1,69 @@
...
+def text():
+    return "Hello, world!"
+
+
+@pytest.fixture
+def text_batch():
+    return ["Hello, world!", "This is a test."]
+
+
+def test_lca_identity_abstraction(transformer, text_batch):
```
Comment (VRehnberg) [inline tests/test_abstractions.py:38]: We could consider a mark for tests that has transformer_lens dependencies and move the import inside the `def transformer():` function to be able to run the rest of tests without the extra dependency. But, it's okay as is.

Comment (VRehnberg) [reply tests/test_abstractions.py:38]: Oh, but all tests in this file has that dependency. Then no mark is needed.

**Review by VRehnberg — COMMENTED** [review]
Looks great in general. Cleans up a lot of things and very well documented.

I made a few comments where I saw possible improvements but nothing really critical. Take a look at the comments and feel free to merge when you're happy with them.

I haven't run the test-suite. I can do that tomorrow, if you'd like.

File: src/cupbearer/detectors/abstraction/abstraction.py [hunk 10]
```diff
@@ -1,198 +1,231 @@
...
-        return False, None
 
-    for vertex in graph:
-        found_cycle, cycle = visit(vertex)
-        if found_cycle:
-            raise ValueError(f"Cycle found in activation graph: {cycle}")
+class _ModuleDict(nn.Module):
+    """A ModuleDict that allows '.' in keys (but not '/').
 
+    Pytorch's ModuleDict doesn't allow '.' in keys, but our activation names contain
```
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/cupbearer/detectors/abstraction/abstraction.py:39]: The dots come from pytorch, not from us (that's how it names nested modules). I think from an end user perspective, you should be able to just pass in the names that e.g. `Module.named_modules()` gives you, rather than having to pass in names with slashes. So the main point of this dict is to hide these issues from users

File: src/cupbearer/detectors/abstraction/abstraction.py [hunk 11]
```diff
@@ -1,198 +1,231 @@
...
+        # BUG: This probably still doesn't behave like a real dict_keys object.
+        # For example, if the underlying dictionary changes, this won't reflect that.
+        return {key.replace("/", "."): None for key in self.dict}.keys()
+
+
+def l2_loss(input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+    assert input.shape == target.shape
+    input = input.reshape(input.size(0), -1)
+    target = target.reshape(target.size(0), -1)
+    # Not totally clear whether mean or sum makes more sense here.
```
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/cupbearer/detectors/abstraction/abstraction.py:87]: I don't think that's a super strong reason tbh, but I do agree mean is probably better, since if we aggregate over different activations with different hidden dims, we probably don't want larger activations to dominate. But I think it's not clear whether just taking a mean here is a principled way of solving that

File: src/cupbearer/detectors/activation_based.py (hunk 2 above)
Comment (ejnnr) [reply src/cupbearer/detectors/activation_based.py]: I've just removed the checksum entirely. There's a similar problem with `activation_preprocessing_func` anyway. I guess we could look at the code for the function and use that, but that won't easily work for lambdas, overall it just seems way too complex. (And I don't just want to use the `id` because that won't work for loading a cache from disk.) So better to warn users that it's their responsibility

File: src/cupbearer/detectors/statistical/statistical.py (hunk 5 above)
Comment (ejnnr) [reply src/cupbearer/detectors/statistical/statistical.py:126]: Yes, same thing. Personally, I really like einops (though this case is simple enough that either seems fine)

File: src/cupbearer/detectors/statistical/statistical.py (hunk 6 above)
Comment (ejnnr) [reply src/cupbearer/detectors/statistical/statistical.py:135]: Yep

File: src/cupbearer/utils/__init__.py (hunk 8 above)
Comment (ejnnr) [reply src/cupbearer/utils/__init__.py:42]: Yeah, I'm not doing it consistently yet, but should probably get used to it as more people start working on the codebase
